Feature/#8693r5d5g add basic console menu 2

Sokol3783 opened the pull request.

Sokol3783 committed:
update pom

Sokol3783 committed:
add BookEntity

Sokol3783 committed:
add reader entity

Sokol3783 committed:
add BaseConsoleMenu

Sokol3783 committed:
remove old test and run menu from entrypoint

Sokol3783 committed:
add test for successful build of app

Sokol3783 committed:
add test methods for base menu

Sokol3783 committed:
write test for print welcome message, print books and print readers

Sokol3783 committed:
add test for exit command and incorect input

Sokol3783 committed:
setInt doesn't add data to scanner so test's didn't pass

Sokol3783 committed:
fix tests for correct working

Sokol3783 committed:
add message for invalid input

ahonsu commented on src/main/java/org/example/ui/BaseConsoleMenu.java:
```diff
@@ -0,0 +1,81 @@
+
+    public BaseConsoleMenu() {
+        scanner = new Scanner(System.in);
+        welcomeMessage = "WELCOME TO THE LIBRARY!";
```
Нет смысла инициализировать это значение в конструкторе. Спокойно можно сделать эту переменную константой.

Sokol3783 replied:
сканнер имеет смысл, пока переписывал тесты помучался, но вроде разобрался, получается если  постоянно не переиницилизировать сканнер то инпут стрим не хочет обновляться, каждый раз для корректного прохождения теста нужно заново кормить данные через system.setin. Видел экзотичные варианты через Process и не стал пробовать, это слишком как по мне.

ahonsu replied:
Теперь понял. Я не подумал, что написание тестов было одним из аргументов.

Альтернативным решением было бы добавление сеттера для объекта сканнера. Через сеттер ты мог бы в своих тестах сетить свежий объект сколько угодно раз.

ahonsu commented on src/main/java/org/example/ui/BaseConsoleMenu.java:
```diff
@@ -0,0 +1,81 @@
+                        case "2" : printReaders();
+                        default  : printInvalidOption();
+                    }
+                }
```
1. Можно ввод пользователя сходу трансформировать в lower/upper case и обрабатывать все пункты меню вместе с exit
2. Сценарий выхода из программы тоже включить в тот же switch-case
3. Switch-case написан на устаревшем синтаксисе, нужно заменить на лямбды
4. Switch-case написан не корректно - если я выбираю пункт 1, по факту программа исполнит все 3 кейса, включая "не валидную опцию"

ahonsu commented on src/main/java/org/example/ui/BaseConsoleMenu.java:
```diff
@@ -0,0 +1,81 @@
+                    }
+                }
+            }
+            System.out.println(getTextMenu());
```
При старте программы не видно основного меню. Т.е. ты показываешь приветствие и сразу же ожидаешь ввода команды.
Нужно скорректировать порядок действий.

Sokol3783 replied:
Исправил

ahonsu commented on src/main/java/org/example/ui/BaseConsoleMenu.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private void printInvalidOption() {
+        System.out.println("Invalid option!");
```
Как вариант: тут можно использовать `System.err.println()` - это подсветит вывод красным цветом, как бы намекая, что возникла ошибка.

ahonsu commented on src/main/java/org/example/ui/BaseConsoleMenu.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private void printReaders() {
+        String buffer = String.format(" %2s %27s\n", "id", "name") +
+                getReaders().stream().sorted(Comparator.comparing(ReaderEntity::getId)).map(
+                                s -> String.format(" %4d %-25s", s.getId(), s.getName()))
+                        .collect(Collectors.joining("\n"));
+        System.out.println(buffer);
```
Я предлагаю упростить вывод в консоль.
Ты все написал корректно и хорошо используешь форматтер для строк.
Проблема - у тебя строка хедера не "адаптируется" под реальную длину имен и названий (для книг такое же замечание). Т.е. если я сделаю очень длинное имя автора, например, то отступ хедера для имен автора у тебя останется фиксированным и "верстка" нарушится.

Понятное дело, это можно решить, написав более хитрую логику вывода в консоль.
Это не цель нашего проекта. Предлагаю просто убрать строку хедера вообще и при выводе в консоль добавить названия полей, типа `ID = 3, author = ... , title = ...`

Дополнительно это поволит обойтись без `String buffer` - можно будет печатать лист объектов не покидая стрим, при помощи `.forEach()`

ahonsu replied:
Да, теперь выглядит проще - и вывод в консоль, и код.

Два пожелания:
1. Добавь пожалуйста какие-то разделители (зяпятую, точку с зяпятой, пайп...) между полями:
![image](https://github.com/Sokol3783/book-library/assets/12547899/41abacbe-6945-4042-818d-1b3994e244f7)
2. Ты формируешь строку для вывода в консоль прямо внутри стрима, есть смысл инкапсулировать это внутри самого класса, как `toString()` метод

ahonsu commented on src/main/java/org/example/entity/BookEntity.java:
```diff
@@ -0,0 +1,40 @@
+package org.example.entity;
+
+public class BookEntity {
```
Предлагаю убрать слово `Entity` из названий этих классов.
На данный момет это не приносит никакой пользы.

Sokol3783 committed:
interrupt threads after finishing testing

Sokol3783 committed:
rename methods of test

Sokol3783 committed:
remove redundant throws

Sokol3783 committed:
adding logging using System.err for looking bugs in tests

Sokol3783 committed:
add test for correct behavior of console menu

Sokol3783 committed:
add logging output of console menu and write print pattern

Sokol3783 committed:
print option on start

Sokol3783 committed:
forget to add run on startup

Sokol3783 committed:
remove unused function

Sokol3783 committed:
add implementation of test for output data of books and readers

Sokol3783 committed:
remove useless variable

Sokol3783 committed:
implement tet of valid working of switch case

Sokol3783 committed:
refactor print reader name

Sokol3783 committed:
refactor print books

Sokol3783 committed:
write tests woke only if run single

Sokol3783 committed:
write working instance

Sokol3783 committed:
remove from class names "Entity"

Sokol3783 committed:
refactor input to lower case

Sokol3783 committed:
make welcome message and scanner static

Sokol3783 committed:
rename methods

Sokol3783 committed:
disable all test witch broke runtime of another tests
broke when scanner read - "exit", after next test work incorrect.

also remove useless function of previous try to write correct working tests.

Sokol3783 committed:
add temp logging for tests for understanding what happen in runtime.

also close scanner after stopping work

Sokol3783 committed:
rename constant

Sokol3783 committed:
remove logging

Sokol3783 committed:
override toString()

Sokol3783 committed:
remove static modifier of terminate and delete @Disabled, all tests work
